misc/python/bin/mzcompose.py

```diff
@@ -0,0 +1,151 @@
+
+    # Determine what images this particular compose file depends upon.
+    say("==> Collecting mzbuild dependencies")
+    images = []
+    with open(config_file) as f:
+        compose = yaml.safe_load(f)
+        for config in compose["services"].values():
+            if "mzbuild" in config:
+                image_name = config["mzbuild"]
+                if image_name not in repo.images:
+                    print("mzcompose: unknown image {}".format(image_name),
+                          file=sys.stderr)
+                    return 1
+                image = repo.images[image_name]
+                config["image"] = image.spec()
+                del (config["mzbuild"])
+                if "propagate-uid-gid" in config:
+                    config["user"] = "{}:{}".format(os.getuid(), os.getgid())
+                    del (config["propagate-uid-gid"])
+                images.append(image)
+
+    deps = repo.resolve_dependencies(images)
+    for d in deps:
+        say(d.spec())
+
+    # Check if the command is going to create or start containers, and if so
+    # build the dependencies. This can be slow, so we don't want to do it if we
+    # can help it (e.g., for `down` or `ps`).
+    if args.command in ["create", "run", "start", "up"]:
+        deps.acquire()
+
+    # Construct a configuration that will point Docker Compose at the correct
+    # images.
+    tempfile = TemporaryFile()
+    os.set_inheritable(tempfile.fileno(), True)
+    yaml.dump(compose, tempfile, encoding="utf-8")  # type: ignore
+    tempfile.flush()
+    tempfile.seek(0)
+
+    # Hand over control to Docker Compose.
+    say("==> Delegating to Docker Compose")
+    dc_args = [
+        "docker-compose",
+        "-f",
+        "/dev/fd/{}".format(tempfile.fileno()),
+        "--project-directory",
+        args.project_directory or str(Path(config_file).parent),
+        *unknown_args,
+        *([args.command] if args.command is not None else []),
+        *args.extra,
+    ]
+    os.execvp("docker-compose", dc_args)
+
+
+def gen_shortcuts(repo: mzbuild.Repository) -> int:
+    template = """#!/usr/bin/env bash
+
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+#
+# mzcompose — runs Docker Compose with Materialize customizations.
+
+exec "$(dirname "$0")/{}/bin/mzcompose" "$@"
+"""
+    for path in repo.compose_dirs:
+        mzcompose_path = path / "mzcompose"
+        with open(str(mzcompose_path), "w") as f:
+            f.write(template.format(os.path.relpath(str(repo.root),
+                                                    str(path))))
+        mzbuild.chmod_x(mzcompose_path)
+    return 0
+
+
+class ArgumentParser(argparse.ArgumentParser):
```

Comment:
can you add a comment explaining why this is necessary?

Reply:
Yep, done.